python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java

```diff
@@ -45,37 +48,94 @@
 
   private final FileLinesContextFactory fileLinesContextFactory;
 
+  private boolean seenFirstToken;
+
+  private final boolean enableNoSonar;
+
+  private final boolean ignoreHeaderComments;
+
+  private Set<Integer> noSonar = Sets.newHashSet();
   private Set<Integer> linesOfCode = Sets.newHashSet();
   private Set<Integer> linesOfComments = Sets.newHashSet();
+  private Set<Integer> linesOfDocstring = Sets.newHashSet();
   private final FileSystem fileSystem;
   private final Map<InputFile, Set<Integer>> allLinesOfCode;
 
-  public FileLinesVisitor(FileLinesContextFactory fileLinesContextFactory, FileSystem fileSystem, Map<InputFile, Set<Integer>> linesOfCode) {
+  public FileLinesVisitor(
+      FileLinesContextFactory fileLinesContextFactory,
+      FileSystem fileSystem,
+      Map<InputFile, Set<Integer>> linesOfCode,
+      boolean enableNoSonar,
+      boolean ignoreHeaderComments) {
     this.fileLinesContextFactory = fileLinesContextFactory;
     this.fileSystem = fileSystem;
     this.allLinesOfCode = linesOfCode;
+    this.enableNoSonar = enableNoSonar;
+    this.ignoreHeaderComments = ignoreHeaderComments;
+  }
+
+  @Override
+  public void init() {
+    DocstringExtractor.DOCUMENTABLE_NODE_TYPES.stream().forEach(this::subscribeTo);
   }
 
   @Override
+  public void visitFile(AstNode astNode) {
+    noSonar = new HashSet<>();
+    linesOfComments = new HashSet<>();
+    seenFirstToken = false;
+  }
+
+  @Override
+  public void visitNode(AstNode astNode) {
+    Token docstringToken = DocstringExtractor.extractDocstring(astNode);
+    if (docstringToken != null) {
+      TokenLocation location = new TokenLocation(docstringToken);
+      for (int line = location.startLine(); line <= location.endLine(); line++) {
+        linesOfDocstring.add(line);
+      }
+    }
+  }
+
+  /**
+   * Gets the lines of codes and lines of comments (with character #).
+   * Does not get the lines of docstrings.
+   */
+  @Override
   public void visitToken(Token token) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this method to reduce its Cognitive Complexity from 24 to the 15 allowed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)